lightning/src/ln/channelmanager.rs

```diff
@@ -2731,6 +2738,43 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		}
 	}
 
+	/// Send a payment that is probing the given route for liquidity. We calculate the
+	/// [`PaymentHash`] of probes based on a static secret and a random [`PaymentId`], which allows
+	/// us to easily discern them from real payments. This can be checked by calling
+	/// [`payment_is_probe`].
+	///
+	/// [`payment_is_probe`]: Self::payment_is_probe
+	pub fn send_probe_payment(&self, route: &Route) -> Result<(PaymentHash, PaymentId), PaymentSendFailure> {
+		let payment_id = PaymentId(self.keys_manager.get_secure_random_bytes());
+
+		let mut preimage = [0u8; 64];
+		preimage[..32].copy_from_slice(&self.probing_cookie_secret);
+		preimage[32..].copy_from_slice(&payment_id.0);
+
+		let payment_hash = PaymentHash(Sha256::hash(&preimage).into_inner());
+
+		if route.paths.len() > 1 {
+			return Err(PaymentSendFailure::ParameterError(APIError::APIMisuseError {
+				err: "Probe payments need to be sent over a single path".to_string()
+			}))
+		}
+		
+		match self.send_payment_internal(route, payment_hash, &None, None, Some(payment_id), None) {
+			Ok(payment_id) => Ok((payment_hash, payment_id)),
+			Err(e) => Err(e)
+		}
+	}
+
+	/// Returns whether a payment with the given [`PaymentHash`] and [`PaymentId`] is, in fact, a
+	/// payment probe.
+	pub fn payment_is_probe(&self, payment_hash: PaymentHash, payment_id: PaymentId) -> bool {
+		let mut preimage = [0u8; 64];
+		preimage[..32].copy_from_slice(&self.probing_cookie_secret);
+		preimage[32..].copy_from_slice(&payment_id.0);
+		let target_payment_hash = PaymentHash(Sha256::hash(&preimage).into_inner());
+		target_payment_hash == payment_hash
+	}
```

Comment:
Hmm... I wonder if it would simplify things if we generate separate `ProbeSuccessful` and `ProbeFailed` events instead. It would simplify the `InvoicePayer` and provide users an indication of the probe's outcome, even if updating the scorer is handled for them.

Reply:
Hum, I agree that detecting earlier that a returning payment is a probe and then yielding corresponding `ProbeSuccessful`/`ProbeFailed` events would be a cleaner solution. However, not sure if it would simplify much since we'd still want to handle these events in `InvoicePayer`, since this is the only place where we have access to all the necessary modules?

Reply:
Yeah, it saves the back-and-forth with `Payer`, thus removing a method from the interface. `InvoicePayer` would still handle the probe, but it would not swallow it.